### PR TITLE
Feature/UI

### LINKED ISSUE #5: Basic UI with GTK and RELM4

### PR DESCRIPTION
**Pull request by eneoli**
This PR adds basic screenshot functionality to flakeshot. Including:

- basic UI showing screenshots
- ~~Selection dialog for selecting croppings of screenshots~~ - will be done in next iteration
- basic UI to save/copy to clipboard the selection

Closes #5

### CODE REVIEW

**Comment by eneoli** [comment]
We will implement basic cropping in the next iteration as this MR is already quite large and we can distribute work better this way.

@TornaxO7  This PR is blocked until #58 is resolved.

**Comment by eneoli** [comment]
Things done:

- added ui (fullscreen windows that show screenshots)
- added toolbar
- added file chooser
- replaced anyhow with thiserror in wayland backend

File: src/frontend/ui_manager.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        FileChooser::open(move |file| {
+            if let Some(path) = file {
+                let width = canvas_ref.borrow().width() as u32;
+                let height = canvas_ref.borrow().height() as u32;
+                canvas_ref
+                    .borrow()
+                    .crop_to_image(0.0, 0.0, width, height)
+                    .expect("Couldn't crop canvas.")
+                    .save(path)
+                    .expect("Couldn't save image.");
```
Comment (TornaxO7) [inline src/frontend/ui_manager.rs:73]: maybe make use of anyhows `with_context` here instead of `expect`?

File: src/frontend/ui/canvas.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        height: u32,
+    ) -> anyhow::Result<DynamicImage> {
+        let mut output_surface = self.crop(x, y, width as i32, height as i32)?;
+        let output_data = output_surface.data()?;
+
+        let mut data_vec = output_data.to_vec();
+        Self::invert_rgba_vec(&mut data_vec);
+
+        let img = RgbaImage::from_vec(width, height, data_vec)
+            .expect("Couldn't create image from buffer");
```
Comment (TornaxO7) [inline src/frontend/ui/canvas.rs:84]: maybe using anyhows `with_context` instead of `expect`

File: src/frontend/main_window.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        _root: &Self::Root,
+        sender: ComponentSender<Self>,
+    ) -> relm4::ComponentParts<Self> {
+        let app = relm4::main_application();
+        let sender_ref = Rc::new(sender);
+
+        let mut monitors = get_monitors();
+        let (total_width, total_height) = get_total_view_size(&monitors.values().collect());
+
+        let mut model = Self::init(total_width, total_height);
```
Comment (TornaxO7) [inline src/frontend/main_window.rs:60]:
```suggestion
        let mut model = {
            let (total_width, total_height) = get_total_view_size(&monitors.values().collect());
            Self::init(total_width, total_height)
        };
```

**Review by TornaxO7 — CHANGES_REQUESTED** [review]
In general, I'd, if it's possible, use anyhows `with context` because it's error output is better formatted than from `expected`.

File: src/frontend/file_chooser.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+#[derive(Debug)]
+pub enum FileChooserEvent {
+    Cancel,
+    Save(String),
+}
+
+pub struct FileChooser {
+    on_submit: Box<dyn Fn(Option<String>) -> ()>,
```
[the remark this answers is not in the document]
Comment (eneoli) [reply src/frontend/file_chooser.rs:20]: We could but why should we?

File: src/frontend/ui/canvas.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            let g = data[i + 1];
+            let b = data[i + 2];
+            let a = data[i + 3];
+
+            data[i] = b;
+            data[i + 1] = g;
+            data[i + 2] = r;
+            data[i + 3] = a;
+        }
+    }
```
Comment (TornaxO7) [inline src/frontend/ui/canvas.rs:101]: I'd actually replace this part with something similar to [this](https://github.com/eneoli/flakeshot/blob/main/src/backend/x11.rs#L134-L139).

**Review by TornaxO7 — CHANGES_REQUESTED** [review]

**Review by TornaxO7 — CHANGES_REQUESTED** [review]
One last thing

**Review by TornaxO7 — APPROVED** [review]

**Pull request merged**